src/statsd.zig

```diff
@@ -0,0 +1,85 @@
+const std = @import("std");
+
+const IO = @import("io.zig").IO;
+const FIFO = @import("fifo.zig").FIFO;
+
+const BufferCompletion = struct {
+    next: ?*BufferCompletion = null,
+    buffer: [256]u8,
+    completion: IO.Completion = undefined,
+};
+
+pub const StatsD = struct {
+    socket: std.os.socket_t,
+    io: *IO,
+    buffer_completions: []BufferCompletion,
+    buffer_completions_fifo: FIFO(BufferCompletion) = .{ .name = "statsd" },
+
+    /// Creates a statsd instance, which will send UDP packets via the IO instance provided.
+    pub fn init(allocator: std.mem.Allocator, io: *IO, address: std.net.Address) !StatsD {
+        const socket = try io.open_socket(
+            address.any.family,
+            std.os.SOCK.DGRAM,
+            std.os.IPPROTO.UDP,
+        );
+        errdefer std.os.closeSocket(socket);
+
+        const buffer_completions = try allocator.alloc(BufferCompletion, 256);
+        errdefer allocator.free(buffer_completions);
+
+        var statsd = StatsD{
+            .socket = socket,
+            .io = io,
+            .buffer_completions = buffer_completions,
+        };
+
+        for (buffer_completions) |*buffer_completion| {
+            buffer_completion.next = null;
+            statsd.buffer_completions_fifo.push(buffer_completion);
+        }
+
+        // 'Connect' the UDP socket, so we can just send() to it normally.
+        try std.os.connect(socket, &address.any, address.getOsSockLen());
```

Comment:
darwin/windows have `open_socket` set `SOCK_NONBLOCK`. For UDP sockets, does this just bind it to the address for future `send()`s and never return `error.WouldBlock`?

Reply:
Good question - I know on Linux, it just binds it to the address and I believe it does the same on Darwin, both should never block. On Windows, https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-connect seems to suggest that we'll get the same behaviour:

> For a connectionless socket (for example, type SOCK_DGRAM), the operation performed by connect is merely to establish a default destination address that can be used on subsequent [send](https://learn.microsoft.com/en-us/windows/desktop/api/winsock2/nf-winsock2-send)/ [WSASend](https://learn.microsoft.com/en-us/windows/desktop/api/winsock2/nf-winsock2-wsasend) and [recv](https://learn.microsoft.com/en-us/windows/desktop/api/winsock/nf-winsock-recv)/ [WSARecv](https://learn.microsoft.com/en-us/windows/desktop/api/winsock2/nf-winsock2-wsarecv) calls